<?xml version="1.0"?>
<launch>
	<arg name="robot" default="localhost" doc="the hostname of the turtlebot, localhost for simulation"/>
	<group if="$(eval robot == 'localhost')">
		<machine name="turtlebot" address="$(arg robot)" env-loader="/home/ubuntu/loader.sh"/>
	</group>
	<group unless="$(eval robot == 'localhost')">
		<machine name="turtlebot" address="$(arg robot)" user="ubuntu" env-loader="/home/ubuntu/loader.sh"/>
	</group>
	<param name="odom_frame_id" value="odom"/>
	<param name="body_frame_id" value="base_footprint"/>
	<param name="robot_description" command="$(find xacro)/xacro $(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro" /> 
	
	<!-- q_cov is ordered as [sigma_theta, sigma_x, sigma_y, sigma_xtheta, sigma_ytheta, sigma_xy] -->
	<rosparam param="q_cov">[0.01, 0.03, 0.03, 0.002, 0.002, 0.002]</rosparam>
	<!-- r_cov is ordered as [sigma_x, sigma_y, sigma_xy] -->
	<rosparam param="r_cov">[0.01, 0.01, 0.001]</rosparam>
	
	<group if="$(eval robot == 'localhost')">
		<include file="$(find nurtlesim)/launch/start_sim.launch">
			<arg name="use_rviz" value="false"/>
		</include>
	</group>
	
	<group unless="$(eval robot == 'localhost')">
		<param name="left_wheel_joint" value="wheel_left_joint"/>
		<param name="right_wheel_joint" value="wheel_right_joint"/>
		<rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml" />
		<include file="$(find nuturtle_robot)/launch/basic_remote.launch">
			<arg name="robot" value="$(arg robot)"/>
		</include>
		<node machine="turtlebot" pkg="nuturtle_robot" type="turtle_interface" name="turtle_interface"/>
		<node pkg="turtlebot3_teleop" type="turtlebot3_teleop_key" name="turtlebot3_teleop_keyboard" output="screen"/>
	</group>
	
	<node machine="turtlebot" pkg="nuslam" type="slam" name="slam"/>
	<node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" /> 
	<node name="rviz" pkg="rviz" type="rviz" args="-d '$(find nuslam)/config/sim_slam.rviz'"/>
</launch>
